storage/innobase/handler/ha_innodb.cc

```diff
@@ -3633,6 +3644,11 @@ static ulonglong innodb_prepare_commit_versioned(THD* thd, ulonglong *trx_id)
 	if (const trx_t* trx = thd_to_trx(thd)) {
 		*trx_id = trx->id;
 
+		if (trx->id && thd_alter_add_system_versioning(thd) && trx->mod_tables.size() == 1
+		    && trx->mod_tables.begin()->second.is_trx_versioned()) {
+			return get_new_trx_id_locked();
+		}
```

Comment:
This check is worse than just checking bool flag.

Reply:
Well, probably. Is it just an opinion or I should fix it somehow?

Reply:
Update `mod_tables` such way that there would be no need to add this check.